unified_planning/model/agent.py

```diff
@@ -0,0 +1,142 @@
+# Copyright 2021 AIPlan4EU project
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import unified_planning
+from unified_planning.shortcuts import *
+import unified_planning.model.operators as op
+from unified_planning.exceptions import (
+    UPProblemDefinitionError,
+    UPTypeError,
+    UPValueError,
+    UPExpressionDefinitionError,
+)
+from unified_planning.model.walkers import OperatorsExtractor
+from fractions import Fraction
+from typing import List, Dict, Set, Union, Optional
+
+from unified_planning.model.abstract_problem import AbstractProblem
+from unified_planning.model.mixins import (
+    ActionsSetMixin,
+    FluentsSetMixin,
+    UserTypesSetMixin,
+    AgentsSetMixin,
+)
+
+
+class Agent(
+    ActionsSetMixin,
+    FluentsSetMixin,
+    UserTypesSetMixin,
+    AgentsSetMixin,
+):
+
+    """Represents an Agent."""
+
+    def __init__(
+        self,
+        ID,
+        env: "unified_planning.environment.Environment" = None,
+    ):
+        self._env = unified_planning.environment.get_env(env)
+        UserTypesSetMixin.__init__(self, self.has_name)
+        FluentsSetMixin.__init__(self, self.env, self._add_user_type, self.has_name)
+        ActionsSetMixin.__init__(
+            self, self.env, self._add_user_type_method, self.has_name
+        )
+        AgentsSetMixin.__init__(self, self.env, self._has_name_method)
+        self._ID: str = ID
+        self._goals: List["up.model.fnode.FNode"] = []
```

Comment:
Yes. We move the association agent-goal and agnet-init--state to MAPRoblem.

Reply:
OK. Note however that this would require the "dot-operator" to refer to the fuents inside of the agents. 

E.g. if I have an agent "a" with a private fluent "battery" in problem "p" I need to write something like:
```python
p.add_goal(GE(Dot(a, battery), 10))
```
to signify the expression "a.battery >= 10"

Reply:
We do it as follows, which should be equivalent:
p.set_agent_goal(a1,GE(battery,10))
p.set_agent_goal(a2,GE(battery,12))
Indeed, fluents are not inside agents, but are associated to them in MAProblem, through a suitable structure (Map)

Reply:
Well, having the dot operator is strictly more expressvie than putting goals inside agents, because you can force constarints **between** agents. e.g.

`Equals(Dot(agent1, location), Dot(agent2, location))` meaning that `agent1.position == agent2.position`. This goal is not expressibe if goals are local to agents.

I seem to remember that we need this kind of goals, right?

Reply:
I see the point (which I overlooked before). Yes, I think we need an operator like that. Is there anything similar already in the UP that we can take inspiration from?